controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -176,6 +185,93 @@ public void initializeStreamWriters(final ClientFactory clientFactory,
     }
 
     /**
+     * Method to check retention policy and generate new periodic cuts and/or truncate stream at an existing stream cut.
+     * 
+     * @param scope scope
+     * @param stream stream
+     * @param policy retention policy
+     * @param recordingTime time of recording
+     * @param contextOpt operation context
+     * @return future.
+     */
+    public CompletableFuture<Void> autoRetention(final String scope, final String stream, final RetentionPolicy policy,
+                                                 final long recordingTime, final OperationContext contextOpt) {
+        final OperationContext context = contextOpt == null ? streamMetadataStore.createContext(scope, stream) : contextOpt;
+
+        return streamMetadataStore.getStreamCutsFromRetentionSet(scope, stream, context, executor)
+                .thenCompose(retentionSet -> {
+                    StreamCutRecord latestCut = retentionSet.stream()
+                            .max(Comparator.comparingLong(StreamCutRecord::getRecordingTime)).orElse(null);
+                    return checkGenerateStreamCut(scope, stream, context, policy, latestCut, recordingTime)
+                            .thenCompose(x -> truncate(scope, stream, policy, context, retentionSet, recordingTime));
+                });
+
+    }
+
+    private CompletableFuture<Void> checkGenerateStreamCut(String scope, String stream, OperationContext context,
+                                                           RetentionPolicy policy, StreamCutRecord latestCut, long recordingTime) {
+        switch (policy.getType()) {
+            case TIME:
+                if (latestCut == null || recordingTime - latestCut.getRecordingTime() > Duration.ofMinutes(Config.MINIMUM_RETENTION_FREQUENCY_IN_MINUTES).toMillis()) {
+                    return generateStreamCut(scope, stream, context)
+                            .thenCompose(newRecord ->
+                                    streamMetadataStore.addStreamCutToRetentionSet(scope, stream, newRecord, context, executor));
```

Comment:
To confirm, we keep in the metadata store (zk for the purposes of this discussion) a set of stream cuts per stream. The stream cuts generated are supposed to be persisted, yes?

Reply:
yes, all valid stream cuts are generated. If auto-truncation kicks in, it purges all stream cuts before 
 (including) the truncation point.